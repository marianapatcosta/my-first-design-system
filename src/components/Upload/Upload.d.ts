import React, { ChangeEvent, CSSProperties, MouseEventHandler } from 'react';
export interface UploadProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * Upload's label
     */
    label?: string;
    /**
     * Is Upload disabled?
     */
    disabled?: boolean;
    /**
     * the fileTypes than are allowed to upload
     */
    fileTypes?: string;
    /**
     * is possible to upload more than one file?
     */
    areMultipleFilesAllowed?: boolean;
    /**
     * does upload have preview functionality?
     */
    hasPreview?: boolean;
    /**
     * are there images to preview?
     */
    imagesToPreview?: Array<any>;
    /**
     * are there files to preview?
     */
    filesToPreview?: Array<any>;
    /**
     * Change handler
     */
    onUpload: (event: ChangeEvent<HTMLInputElement>) => void;
    /**
     * Delete preview handler
     */
    onDeletePreview?: MouseEventHandler<HTMLButtonElement>;
}
export declare const Upload: React.FC<UploadProps>;
