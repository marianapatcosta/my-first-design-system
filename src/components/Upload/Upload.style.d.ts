/// <reference types="react" />
export declare const StyledUpload: import("styled-components").StyledComponent<import("react").ForwardRefExoticComponent<import("../Button/Button").ButtonProps & import("react").RefAttributes<HTMLButtonElement>>, any, {}, never>;
export declare const StyledUploadInput: import("styled-components").StyledComponent<"input", any, {}, never>;
export declare const StyledUploadPreview: import("styled-components").StyledComponent<"div", any, {
    areThereFilesToPreview?: boolean | undefined;
}, never>;
export declare const StyledUploadPreviewContainer: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledUploadPreviewImage: import("styled-components").StyledComponent<"img", any, {}, never>;
export declare const StyledUploadPreviewFile: import("styled-components").StyledComponent<"p", any, {}, never>;
export declare const StyledUploadPreviewButton: import("styled-components").StyledComponent<"button", any, {
    areThereFilesToPreview?: boolean | undefined;
}, never>;
export declare const StyledUploadWrapper: import("styled-components").StyledComponent<"div", any, {}, never>;
