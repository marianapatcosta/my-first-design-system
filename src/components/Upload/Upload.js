var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useRef, } from 'react';
import { StyledUpload, StyledUploadWrapper, StyledUploadInput, StyledUploadPreview, StyledUploadPreviewContainer, StyledUploadPreviewFile, StyledUploadPreviewImage, StyledUploadPreviewButton, } from './Upload.style';
export var Upload = function (_a) {
    var label = _a.label, disabled = _a.disabled, fileTypes = _a.fileTypes, areMultipleFilesAllowed = _a.areMultipleFilesAllowed, className = _a.className, hasPreview = _a.hasPreview, imagesToPreview = _a.imagesToPreview, filesToPreview = _a.filesToPreview, onUpload = _a.onUpload, onDeletePreview = _a.onDeletePreview, props = __rest(_a, ["label", "disabled", "fileTypes", "areMultipleFilesAllowed", "className", "hasPreview", "imagesToPreview", "filesToPreview", "onUpload", "onDeletePreview"]);
    var inputRef = useRef(null);
    return (_jsxs(StyledUploadWrapper, { children: [_jsx(StyledUpload, __assign({}, props, { label: label, className: className, disabled: disabled, onClick: function () { return inputRef.current && inputRef.current.click(); } }), void 0),
            _jsx(StyledUploadInput, { "aria-label": 'File upload', ref: inputRef, type: 'file', multiple: areMultipleFilesAllowed, accept: fileTypes, onChange: onUpload }, void 0),
            hasPreview && (!!imagesToPreview || !!filesToPreview) && (_jsxs(StyledUploadPreview, __assign({ areThereFilesToPreview: !!filesToPreview && !!filesToPreview.length }, { children: [!!imagesToPreview &&
                        !!imagesToPreview.length &&
                        imagesToPreview.map(function (image, index) { return (_jsxs(StyledUploadPreviewContainer, { children: [_jsx(StyledUploadPreviewImage, { src: image, alt: "uploaded-preview-" + index }, void 0),
                                _jsx(StyledUploadPreviewButton, { "aria-label": "remove uploaded image " + index, onClick: function () { return !!onDeletePreview && onDeletePreview(index); } }, void 0)] }, "image-preview-" + index)); }),
                    !!filesToPreview &&
                        !!filesToPreview.length &&
                        filesToPreview.map(function (file, index) { return (_jsxs(StyledUploadPreviewContainer, { children: [_jsx(StyledUploadPreviewFile, { children: file.name }, void 0),
                                _jsx(StyledUploadPreviewButton, { areThereFilesToPreview: !!filesToPreview && !!filesToPreview.length, "aria-label": "remove uploaded file " + index, onClick: function () { return !!onDeletePreview && onDeletePreview(index); } }, void 0)] }, "file-preview-" + index)); })] }), void 0))] }, void 0));
};
Upload.defaultProps = {
    id: '',
    className: '',
    style: {},
    label: '',
    disabled: false,
    areMultipleFilesAllowed: false,
    fileTypes: '',
    hasPreview: false,
    imagesToPreview: [],
    filesToPreview: [],
    onUpload: function () { return null; },
    onDeletePreview: function () { return null; },
};
