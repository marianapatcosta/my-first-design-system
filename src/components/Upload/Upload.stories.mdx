import { Meta } from '@storybook/addon-docs/blocks';
import { Upload } from './Upload';

<Meta title='Design System|Upload' component={Upload} />

# Upload

Uplaod component

## Usage

Use Upload component when it is required to upload a file (audio, image, doc, etc). If `hasPreviews`prop is passed, it displays the
`imagesToPreview` or `filedToPreview`, correpsonding to the uploaded files

- Default Upload:
  <Story name='Default'>
    <Upload label='Upload' />
  </Story>

## Other important notes

not applicable.
