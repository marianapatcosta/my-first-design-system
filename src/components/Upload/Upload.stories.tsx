import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Upload, UploadProps } from './Upload';

export default {
  title: 'Upload',
  component: Upload,
  args: {},
} as Meta;

const Template: Story<UploadProps> = (args) => <Upload {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'Upload',
};
