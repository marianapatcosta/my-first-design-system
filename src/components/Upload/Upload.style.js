var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';
import { Button } from '../Button/Button';
import { Close } from '../../assets/icons/';
export var StyledUpload = styled(Button)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  width: 100%;\n  padding: 0.3rem;\n"], ["\n  ", "\n  width: 100%;\n  padding: 0.3rem;\n"])), baseStyles);
export var StyledUploadInput = styled.input(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  visibility: hidden;\n  position: absolute;\n  z-index: -1;\n  width: 0;\n  height: 0;\n"], ["\n  visibility: hidden;\n  position: absolute;\n  z-index: -1;\n  width: 0;\n  height: 0;\n"])));
export var StyledUploadPreview = styled.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  flex: 2;\n  overflow: auto;\n  margin-left: 1.25rem;\n  height: 100%;\n\n  ", "\n"], ["\n  display: flex;\n  flex: 2;\n  overflow: auto;\n  margin-left: 1.25rem;\n  height: 100%;\n\n  ",
    "\n"])), function (_a) {
    var areThereFilesToPreview = _a.areThereFilesToPreview;
    return areThereFilesToPreview &&
        "display: block;\n  margin-left: 0;\n  height: 100%;";
});
export var StyledUploadPreviewContainer = styled.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  position: relative;\n  display: flex;\n  align-items: flex-start;\n"], ["\n  position: relative;\n  display: flex;\n  align-items: flex-start;\n"])));
export var StyledUploadPreviewImage = styled.img(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  width: 6.25rem;\n  height: 4.4rem;\n  margin-right: 0.3rem;\n"], ["\n  width: 6.25rem;\n  height: 4.4rem;\n  margin-right: 0.3rem;\n"])));
export var StyledUploadPreviewFile = styled.p(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  font-size: 90%;\n  display: inline;\n  margin-right: 0.3rem;\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n"], ["\n  font-size: 90%;\n  display: inline;\n  margin-right: 0.3rem;\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n"])));
export var StyledUploadPreviewButton = styled.button(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  border: none;\n  position: absolute;\n  top: 0.3rem;\n  right: 0.8em;\n  width: 0.625rem;\n  height: 0.625rem;\n  padding: 0.063rem;\n  cursor: pointer;\n  border-radius: 50%;\n  -webkit-border-radius: 50%;\n  -moz-border-radius: 50%;\n  -ms-border-radius: 50%;\n  -o-border-radius: 50%;\n  background: ", " url(", ") no-repeat center center;\n  background-size: 0.7rem 0.7rem;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  ", "\n"], ["\n  border: none;\n  position: absolute;\n  top: 0.3rem;\n  right: 0.8em;\n  width: 0.625rem;\n  height: 0.625rem;\n  padding: 0.063rem;\n  cursor: pointer;\n  border-radius: 50%;\n  -webkit-border-radius: 50%;\n  -moz-border-radius: 50%;\n  -ms-border-radius: 50%;\n  -o-border-radius: 50%;\n  background: ", " url(", ") no-repeat center center;\n  background-size: 0.7rem 0.7rem;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  ",
    "\n"])), colors.white, Close, function (_a) {
    var areThereFilesToPreview = _a.areThereFilesToPreview;
    return areThereFilesToPreview &&
        "dposition: relative;\ntop: 0;\nright: 0;\n";
});
export var StyledUploadWrapper = styled.div(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  display: flex;\n  align-items: flex-start;\n  justify-content: space-between;\n  overflow: hidden;\n  height: 5.6rem;\n  width: 100%;\n  max-width: 30rem;\n"], ["\n  display: flex;\n  align-items: flex-start;\n  justify-content: space-between;\n  overflow: hidden;\n  height: 5.6rem;\n  width: 100%;\n  max-width: 30rem;\n"])));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8;
