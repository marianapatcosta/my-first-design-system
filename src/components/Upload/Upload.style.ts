import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';
import { Button } from '../Button/Button';
import { Close } from '../../assets/icons/';

export const StyledUpload = styled(Button)`
  ${baseStyles}
  width: 100%;
  padding: 0.3rem;
`;

export const StyledUploadInput = styled.input`
  visibility: hidden;
  position: absolute;
  z-index: -1;
  width: 0;
  height: 0;
`;

export const StyledUploadPreview = styled.div<{
  areThereFilesToPreview?: boolean;
}>`
  display: flex;
  flex: 2;
  overflow: auto;
  margin-left: 1.25rem;
  height: 100%;

  ${({ areThereFilesToPreview }) =>
    areThereFilesToPreview &&
    `display: block;
  margin-left: 0;
  height: 100%;`}
`;

export const StyledUploadPreviewContainer = styled.div`
  position: relative;
  display: flex;
  align-items: flex-start;
`;

export const StyledUploadPreviewImage = styled.img`
  width: 6.25rem;
  height: 4.4rem;
  margin-right: 0.3rem;
`;

export const StyledUploadPreviewFile = styled.p`
  font-size: 90%;
  display: inline;
  margin-right: 0.3rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const StyledUploadPreviewButton = styled.button<{
  areThereFilesToPreview?: boolean;
}>`
  border: none;
  position: absolute;
  top: 0.3rem;
  right: 0.8em;
  width: 0.625rem;
  height: 0.625rem;
  padding: 0.063rem;
  cursor: pointer;
  border-radius: 50%;
  -webkit-border-radius: 50%;
  -moz-border-radius: 50%;
  -ms-border-radius: 50%;
  -o-border-radius: 50%;
  background: ${colors.white} url(${Close}) no-repeat center center;
  background-size: 0.7rem 0.7rem;

  :hover {
    opacity: 0.75;
  }

  ${({ areThereFilesToPreview }) =>
    areThereFilesToPreview &&
    `dposition: relative;
top: 0;
right: 0;
`}
`;

export const StyledUploadWrapper = styled.div`
  display: flex;
  align-items: flex-start;
  justify-content: space-between;
  overflow: hidden;
  height: 5.6rem;
  width: 100%;
  max-width: 30rem;
`;
