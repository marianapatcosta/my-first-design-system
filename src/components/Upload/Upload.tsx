import React, {
  ChangeEvent,
  CSSProperties,
  MouseEventHandler,
  useRef,
} from 'react';
import {
  StyledUpload,
  StyledUploadWrapper,
  StyledUploadInput,
  StyledUploadPreview,
  StyledUploadPreviewContainer,
  StyledUploadPreviewFile,
  StyledUploadPreviewImage,
  StyledUploadPreviewButton,
} from './Upload.style';

export interface UploadProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * Upload's label
   */
  label?: string;
  /**
   * Is Upload disabled?
   */
  disabled?: boolean;
  /**
   * the fileTypes than are allowed to upload
   */
  fileTypes?: string;
  /**
   * is possible to upload more than one file?
   */
  areMultipleFilesAllowed?: boolean;
  /**
   * does upload have preview functionality?
   */
  hasPreview?: boolean;
  /**
   * are there images to preview?
   */
  imagesToPreview?: Array<any>;
  /**
   * are there files to preview?
   */
  filesToPreview?: Array<any>;
  /**
   * Change handler
   */
  onUpload: (event: ChangeEvent<HTMLInputElement>) => void;
  /**
   * Delete preview handler
   */
  onDeletePreview?: MouseEventHandler<HTMLButtonElement>;
}

export const Upload: React.FC<UploadProps> = ({
  label,
  disabled,
  fileTypes,
  areMultipleFilesAllowed,
  onUpload,
  className,
  hasPreview,
  imagesToPreview,
  filesToPreview,
  onDeletePreview,
  ...props
}) => {
  const inputRef = useRef<HTMLInputElement>(null);

  return (
    <StyledUploadWrapper>
      <StyledUpload
        {...props}
        label={label}
        className={className}
        disabled={disabled}
        onClick={() => inputRef.current && inputRef.current.click()}
      />
      <StyledUploadInput
        aria-label='File upload'
        ref={inputRef}
        type='file'
        multiple={areMultipleFilesAllowed}
        accept={fileTypes}
        onChange={onUpload}
      />
      {hasPreview && (!!imagesToPreview || !!filesToPreview) && (
        <StyledUploadPreview
          areThereFilesToPreview={!!filesToPreview && !!filesToPreview.length}
        >
          {!!imagesToPreview &&
            !!imagesToPreview.length &&
            imagesToPreview.map((image: any, index: any) => (
              <StyledUploadPreviewContainer key={`image-preview-${index}`}>
                <StyledUploadPreviewImage
                  src={image}
                  alt={`uploaded-preview-${index}`}
                />
                <StyledUploadPreviewButton
                  aria-label={`remove uploaded image ${index}`}
                  onClick={() => !!onDeletePreview && onDeletePreview(index)}
                />
              </StyledUploadPreviewContainer>
            ))}
          {!!filesToPreview &&
            !!filesToPreview.length &&
            filesToPreview.map((file: any, index: any) => (
              <StyledUploadPreviewContainer key={`file-preview-${index}`}>
                <StyledUploadPreviewFile>{file.name}</StyledUploadPreviewFile>
                <StyledUploadPreviewButton
                  areThereFilesToPreview={
                    !!filesToPreview && !!filesToPreview.length
                  }
                  aria-label={`remove uploaded file ${index}`}
                  onClick={() => !!onDeletePreview && onDeletePreview(index)}
                />
              </StyledUploadPreviewContainer>
            ))}
        </StyledUploadPreview>
      )}
    </StyledUploadWrapper>
  );
};

Upload.defaultProps = {
  id: '',
  className: '',
  style: {},
  label: '',
  disabled: false,
  areMultipleFilesAllowed: false,
  fileTypes: '',
  hasPreview: false,
  imagesToPreview: [],
  filesToPreview: [],
  onUpload: () => null,
  onDeletePreview: () => null,
};
