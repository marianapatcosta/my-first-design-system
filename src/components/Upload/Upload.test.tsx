import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Upload } from './Upload';

describe('Upload', () => {
  const defaultProps = { onUpload: () => null, label: 'Upload' };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Upload {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display label', () => {
    const { getByText } = render(<Upload {...defaultProps} />);
    getByText('Upload');
  });

  test('should handle change events', () => {
    const onUploadMock = jest.fn();
    const { getByLabelText } = render(
      <Upload {...defaultProps} onUpload={onUploadMock} />
    );

    const input = getByLabelText('File upload');

    fireEvent.change(input);

    expect(onUploadMock).toBeCalled();
  });

  test('should return null on click if onUpload prop is not passed', () => {
    const result =
      Upload.defaultProps?.onUpload && Upload.defaultProps.onUpload();
    expect(result).toBeNull();
  });

  test('should display images to preview if `hasPreview`prop is passed', () => {
    const { getByAltText } = render(
      <Upload
        {...defaultProps}
        hasPreview
        imagesToPreview={[{ name: 'image 1', src: '' }]}
      />
    );

    const imageToPreview = getByAltText('uploaded-preview-0');
    expect(imageToPreview).toBeInTheDocument();
  });

  test('should call onDeletePreviews if delete button of images is clicked', () => {
    const onDeleteMock = jest.fn();
    const { getByLabelText } = render(
      <Upload
        {...defaultProps}
        hasPreview
        imagesToPreview={[{ name: 'image 1', file: '' }]}
        onDeletePreview={onDeleteMock}
      />
    );

    const imageToPreviewButton = getByLabelText('remove uploaded image 0');
    fireEvent.click(imageToPreviewButton);
    expect(onDeleteMock).toBeCalled();
  });

  test('should display files to preview if `hasPreview`prop is passed', () => {
    const { getByLabelText } = render(
      <Upload
        {...defaultProps}
        hasPreview
        filesToPreview={[{ name: 'file 1', file: '' }]}
      />
    );

    const fileToPreviewButton = getByLabelText('remove uploaded file 0');
    expect(fileToPreviewButton).toBeInTheDocument();
  });

  test('should call onDeletePreviews if delete button of files is clicked', () => {
    const onDeleteMock = jest.fn();
    const { getByLabelText } = render(
      <Upload
        {...defaultProps}
        hasPreview
        filesToPreview={[{ name: 'file 1', file: '' }]}
        onDeletePreview={onDeleteMock}
      />
    );

    const fileToPreviewButton = getByLabelText('remove uploaded file 0');
    fireEvent.click(fileToPreviewButton);
    expect(onDeleteMock).toBeCalled();
  });

  test('should return null on delete button click if `onDeletePreview` prop is not passed', () => {
    const result =
      Upload.defaultProps?.onDeletePreview &&
      Upload.defaultProps.onDeletePreview();
    expect(result).toBeNull();
  });
});
