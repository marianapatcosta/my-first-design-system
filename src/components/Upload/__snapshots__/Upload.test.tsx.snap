// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload renders without errors and matches snapshot 1`] = `
<div
  class="sc-jrAGrp igiFOb"
>
  <button
    class="sc-bdfBwQ iRImph sc-hKgILt AmXCn"
    id=""
    type="button"
  >
    
    <span
      class="sc-gsTCUz WUuCA"
    >
      Upload
    </span>
  </button>
  <input
    accept=""
    aria-label="File upload"
    class="sc-eCssSg QNthN"
    type="file"
  />
</div>
`;
